### DISCUSSION TITLE
[A-L] (10 giugno 2019) Esercizio 3

### DISCUSSION THREAD

**Discussion by Frisbon** [sapienzastudentsnetwork/progettazione-di-algoritmi#89] [Esami A-L]
Dato un array $A$ ordinato di numeri reali (quindi $A[1] \le A[2] \le \dots \le A[n]$),
dare lo pseudocodice per un algoritmo per trovare un insieme di cardinalità minima di intervalli di lunghezza uno (quindi intervalli di tipo $[x, x+1]$) che contengono tutti gli elementi $A[i]$.
Per esempio, dato $A = [1.1,2,2.05,3,4]$, una soluzione sarebbe { $[1.1 , 2.1]\ ,\ [3,4]$ }.
La complessità dell'algoritmo dovrebbe essere $O(n)$.

> Punti bonus se riesci a decifrare ciò che dice

**Comment by Frisbon** [comment]
dovrebbe funzionare anche con altri esempi
```py
def bruh(A):
    sol = []
    i = 0;
    while i<len(A)-1: #O(n)
        fine_intervallo = A[i]+1
        sol.append( (A[i],A[i]+1) )
        while A[i]<fine_intervallo and i<len(A)-1: 
            i += 1
    return sol

print(bruh([1.1, 2 , 2.05,3, 33, 34,34.1])) 
```
Il costo è $O(n)$ perchè nonostante i due while, comunque visito tutti gli elementi una sola volta (o nel ciclo fuori o in quello dentro)

**Comment by CasuFrost** [comment]
![image](https://github.com/user-attachments/assets/b8c666be-74c8-4f0e-8871-33d597c8cfda)

**Comment by daromacarlo** [comment]
![immagine](https://github.com/user-attachments/assets/5802974f-9cf3-4812-bb1e-883c8f8b7c1b)

**Comment by cek753** [comment]
```

Algoritmo(A):
    sol = {}
    sol.add(A[1],A[1]+1)
    x  = A[1]+1
    for i in len(A):
        if A[i]>x:
            sol.add(A[i],A[i]+1)
            x=A[i]+1
    return sol
    
```
